File: testapp/src/main/java/com/appliedrec/mrtd_reader_app/BACSpecModel.java
package com.appliedrec.mrtd_reader_app;

import android.content.SharedPreferences;

import androidx.annotation.MainThread;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.appliedrec.mrtdreader.BACSpec;
import com.google.gson.Gson;

public class BACSpecModel extends ViewModel implements SharedPreferences.OnSharedPreferenceChangeListener {

    private final MutableLiveData<BACSpec> liveData = new MutableLiveData<>();
    private SharedPreferences preferences;
    private final Gson gson = new Gson();
    private static final String PREF_KEY = "bacSpec";
    private BACSpec bacSpec;

    public BACSpecModel() {
    }

    @MainThread
    public void setSharedPreferences(SharedPreferences preferences) {
        this.preferences = preferences;
        this.preferences.registerOnSharedPreferenceChangeListener(this);
        this.bacSpec = bacSpecFromPreferences();
        liveData.setValue(this.bacSpec);
    }

    public LiveData<BACSpec> getBACSpec() {
        return liveData;
    }

    public void setBACSpec(BACSpec bacSpec) {
        this.bacSpec = bacSpec;
        if (preferences == null) {
            return;
        }
        if (bacSpec == null) {
            preferences.edit().remove(PREF_KEY).apply();
        } else {
            preferences.edit().putString(PREF_KEY, gson.toJson(bacSpec)).apply();
        }
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        if (preferences != null) {
            preferences.unregisterOnSharedPreferenceChangeListener(this);
        }
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (PREF_KEY.equals(key)) {
            this.bacSpec = bacSpecFromPreferences();
            liveData.postValue(this.bacSpec);
        }
    }

    private BACSpec bacSpecFromPreferences() {
        if (preferences == null) {
            return null;
        }
        String bacSpecsString = preferences.getString(PREF_KEY, null);
        if (bacSpecsString == null || bacSpecsString.isEmpty()) {
            return null;
        }
        return gson.fromJson(bacSpecsString, BACSpec.class);
    }
}

File: testapp/src/main/java/com/appliedrec/mrtd_reader_app/BACSpecModel.kt
package com.appliedrec.mrtd_reader_app

import android.content.SharedPreferences
import androidx.annotation.MainThread
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.appliedrec.mrtdreader.BACSpec
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

class BACSpecModel : ViewModel(), SharedPreferences.OnSharedPreferenceChangeListener {
    private val _liveData: MutableLiveData<BACSpec?> = MutableLiveData<BACSpec?>()
    private var preferences: SharedPreferences? = null
    private var _bacSpec: BACSpec? = null

    @MainThread
    fun setSharedPreferences(preferences: SharedPreferences?) {
        this.preferences = preferences
        this.preferences?.registerOnSharedPreferenceChangeListener(this)
        _bacSpec = bacSpecFromPreferences()
        _liveData.setValue(_bacSpec)
    }

    var bacSpec: BACSpec?
        get() = _bacSpec
        set(bacSpec) {
            _bacSpec = bacSpec
            if (preferences == null) {
                return
            }
            if (this._bacSpec == null) {
                preferences!!.edit().remove(PREF_KEY).apply()
            } else {
                preferences!!.edit().putString(PREF_KEY, Json.encodeToString(this._bacSpec)).apply()
            }
            _liveData.postValue(_bacSpec)
        }

    val liveData: LiveData<BACSpec?> = _liveData

    override fun onCleared() {
        super.onCleared()
        preferences?.unregisterOnSharedPreferenceChangeListener(this)
    }

    private fun bacSpecFromPreferences(): BACSpec? {
        val bacSpecsString: String = preferences?.getString(PREF_KEY, null) ?: return null
        if (bacSpecsString.isEmpty()) {
            return null
        }
        return Json.decodeFromString(bacSpecsString)
    }

    companion object {
        private const val PREF_KEY = "bacSpec"
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences?, key: String?) {
        if (PREF_KEY == key) {
            _bacSpec = bacSpecFromPreferences()
            _liveData.postValue(_bacSpec)
        }
    }
}